src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,343 @@
+    assert(idx.length == n)
+    resetIndex(idx)
+    quickSortWithIndex(a, idx, 0, n)
+  }
+
+  def computeFractional(input: Array[Double], fractional: Array[Double]): Int = {
+    val n = input.length
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i)
+      val f = x - x.floor
+      fractional(i) = f
+      totalFractional += f
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+    F
+  }
+
+  def roundWithConstantSum(resized: Array[Double], fractional: Array[Double], index: Array[Int],
+    indexInverse: Array[Int], output: IntConsumer, F: Int, expectedSize: Long, keepLast: Boolean = true) {
+    val n = resized.length
+    assert(fractional.length == n && index.length == n && indexInverse.length == n)
+    sortedIndex(fractional, index)
+
+    var i = 0
+    while (i < n) {
+      indexInverse(index(i)) = i
+      i += 1
+    }
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (indexInverse(i) < F) resized(i).ceil else resized(i).floor
+      if (keepLast || i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+    assert(newSize == expectedSize)
+  }
+
+  def headerBlock(sampleIds: Array[String], nVariants: Long): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
+    emitHeaderBlock(bb, sampleIds, nVariants)
+    bb.result()
+  }
+
+  def emitHeaderBlock(bb: ArrayBuilder[Byte], sampleIds: Array[String], nVariants: Long) {
+    val nSamples = sampleIds.length
+    assert(nVariants <= (1L << 32) - 1, s"Maximum number of variants can export is (2^32 - 1). Found $nVariants.")
+
+    val magicNumbers = Array("b", "g", "e", "n").flatMap(_.getBytes)
+    val flags = 0x01 | (0x02 << 2) | (0x01 << 31)
+    val headerLength = 20
+
+    intToBytesLE(bb, 0) // placeholder for offset
+    intToBytesLE(bb, headerLength)
+    intToBytesLE(bb, nVariants.toUInt.intRep)
+    intToBytesLE(bb, nSamples)
+    bb ++= magicNumbers
+    intToBytesLE(bb, flags)
+
+    intToBytesLE(bb, 0) // placeholder for length of sample ids
+    intToBytesLE(bb, nSamples)
+
+    var i = 0
+    var sampleBlockLength = 8
+    while (i < nSamples) {
+      sampleBlockLength += stringToBytesWithShortLength(bb, sampleIds(i))
+      i += 1
+    }
+
+    val offset = headerLength + sampleBlockLength
+    updateIntToBytesLE(bb, offset, 0)
+    updateIntToBytesLE(bb, sampleBlockLength, 24)
+  }
+
+  def emitVariant(bb: ArrayBuilder[Byte], v: Variant, va: Annotation, view: ArrayGenotypeView,
+    rsidQuery: Option[Querier], varidQuery: Option[Querier], nSamples: Int, nBitsPerProb: Int) {
+    val nAlleles = v.nAlleles
+    require(nAlleles <= 0xffff, s"Maximum number of alleles per variant is ${ 0xffff }. Found ${ v.nAlleles }.")
+
+    val varid = varidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(v.toString)
+    val rsid = rsidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(".")
```

Comment:
God, these expressions make my eyes bleed.

Again, the `Row` stuff needs to go away.  I'd make a `BGENVariantAnnotationsView` like `HTSGenotypeView` that lets you query the varid, rsid, etc. fields.